src/combinatorics/quiver/basic.lean

```diff
@@ -124,8 +136,112 @@ instance empty_quiver (V : Type u) : quiver.{u} (empty V) := ⟨λ a b, pempty
 
 @[simp] lemma empty_arrow {V : Type u} (a b : empty V) : (a ⟶ b) = pempty := rfl
 
-
 /-- A quiver is thin if it has no parallel arrows. -/
 @[reducible] def is_thin (V : Type u) [quiver V] := ∀ (a b : V), subsingleton (a ⟶ b)
 
+section push
+/-!
+### Pushing the quiver structure on `V` along a map `V → W`
+-/
+
+/-- The `quiver` instance obtained by pushing arrows of `V` along the map `σ : V → W` -/
+@[nolint unused_arguments]
+def push {V : Type u} [quiver V] {W : Type u₂} (σ : V → W) := W
+
+instance {V : Type u} [quiver V] {W : Type u₂} (σ : V → W) [h : nonempty W] : nonempty (push σ) := h
+
+namespace push
+
+variables {V : Type*} [quiver V] {W : Type*} (σ : V → W)
+
+/-- The quiver structure obtained by pushing arrows of `V` along the map `σ : V → W` -/
+@[nolint has_nonempty_instance]
+inductive push_quiver {V : Type u} [quiver.{v} V] {W : Type u₂} (σ : V → W) :
+  W → W → Type (max u u₂ v)
+| arrow {X Y : V} (f : X ⟶ Y) : push_quiver (σ X) (σ Y)
+
+instance : quiver (push σ) := ⟨λ X Y, push_quiver σ X Y⟩
+
+/-- The prefunctor induced by pushing arrows via `σ` -/
+def of : prefunctor V (push σ) :=
+{ obj := σ,
+  map := λ X Y f, push_quiver.arrow f}
+
+@[simp] lemma of_obj : ((of σ)).obj = σ := rfl
+
+variables {W' : Type*} [quiver W'] (φ : prefunctor V W') (τ : W → W') (h : ∀ x, φ.obj x = τ (σ x) )
+
+include φ h
+/-- Any map `φ : V → W'` factoring through `τ : W → W'` lifts to a prefunctor from `V` to `W`. -/
+def lift : prefunctor (push σ) W' :=
+{ obj := τ,
+  map := @push_quiver.rec V _ W σ
+    (λ X Y f, τ X ⟶ τ Y)
+    (λ X Y f, by {rw [←h X,←h Y], exact φ.map f}) }
+
+lemma lift_spec_obj : (lift σ φ τ h).obj = τ := rfl
+
+lemma lift_spec_comm : (of σ).comp (lift σ φ τ h) = φ :=
+begin
+  fapply prefunctor.ext,
+  { rintros, simp only [prefunctor.comp_obj], symmetry, exact h X, },
+  { rintros _ _ f, simp only [prefunctor.comp_map],
+    apply eq_of_heq,
+    iterate 2 { apply (cast_heq _ _).trans },
+    symmetry,
+    iterate 2 { apply (eq_rec_heq _ _).trans },
+    refl, },
+end
+
+lemma lift_unique (Φ : prefunctor (push σ) W') (Φ₀ : Φ.obj = τ) (Φcomm : (of σ).comp Φ = φ) :
+  Φ = (lift σ φ τ h) :=
+begin
+  dsimp only [of,lift],
+  fapply prefunctor.ext,
+  { rintros, simp_rw [←Φ₀], },
+  { rintros _ _ ⟨⟩, subst_vars, simp only [prefunctor.comp_map, cast_eq], refl, }
+end
+
+end push
+
+end push
+
+section cast
+/-!
+### Rewriting arrows along equalities of vertices
+-/
```

Comment:
What is this section for?

Reply:
You mean what's the purpose of the section block, or its content?

Reply:
It's an analogue for https://leanprover-community.github.io/mathlib_docs/combinatorics/simple_graph/connectivity.html#simple_graph.walk.copy but for individual edges (simple graphs don't need this on the edges themselves, but it starts to show up for walks)